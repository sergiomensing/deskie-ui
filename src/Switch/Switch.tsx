import {
  forwardRef,
  InputHTMLAttributes,
  PropsWithChildren,
  useImperativeHandle,
  useRef,
  useState,
  useEffect,
  useCallback,
} from 'react';
import styled, { css } from 'styled-components';

type BaseSwitchProps = {
  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
};
export type SwitchProps = BaseSwitchProps & Omit<InputHTMLAttributes<unknown>, keyof BaseSwitchProps | 'type'>;

const StyledSwitch = styled.label<{ disabled: boolean; checked: boolean }>`
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  display: flex;
  align-items: center;
  user-select: none;
  color: ${(props) => (props.disabled ? props.theme.colors.disabledText : props.theme.colors.label)};

  input[type='checkbox'] {
    opacity: 0;
    outline: none;
    position: absolute;
    width: 0;
    height: 0;
    margin: 0;
    padding: 0;
    z-index: -1;
    background-color: transparent;
  }

  .switch {
    display: flex;
    align-items: center;
    position: relative;
  }

  .switch:before {
    content: '';
    width: 36px;
    height: 24px;
    margin-right: 0.75rem;
    border-radius: 2rem;
    display: flex;
    justify-content: center;
    align-items: center;
    background: ${(props) => (props.disabled ? props.theme.colors.accent2 : props.theme.colors.accent3)};
    transition: 0.1s ease;
  }

  .switch:after {
    content: '';
    background-size: contain;
    background-position: center;
    width: 16px;
    height: 16px;
    left: 4px;
    top: 4px;
    position: absolute;
    border: 2px solid transparent;
    background-color: ${(props) => props.theme.colors.background};
    border-radius: 2rem;
    transition: 0.15s ease;
  }

  input:checked ~ .switch::before {
    background: ${(props) => (props.disabled ? props.theme.colors.accent2 : props.theme.colors.primary)};
  }

  input:checked ~ .switch::after {
    transform: translate3d(12px, 0, 0);
  }

  .switch-label {
    margin-top: 2px;
    white-space: nowrap;
  }

  ${(props) =>
    !props.disabled &&
    css`
      &:hover .switch::before {
        background: ${(props) => props.theme.colors.accent7};
      }
      &:hover input:checked ~ .switch::before {
        background-color: ${(props) => props.theme.colors.primaryHover};
      }

      input:focus-visible ~ .switch:before {
        box-shadow: 0 0 0 2px ${(props) => props.theme.colors.inputBorder};
      }
      input:checked:focus-visible ~ .switch::before {
        box-shadow: 0 0 0 2px ${(props) => props.theme.colors.primaryHover};
      }
    `}
`;

const Switch = forwardRef<HTMLInputElement, PropsWithChildren<SwitchProps>>(
  (
    { disabled = false, checked, defaultChecked, children, onChange, className = '', ...rest },
    ref: React.Ref<HTMLInputElement | null>,
  ) => {
    const [selfChecked, setSelfChecked] = useState<boolean>(() => defaultChecked || false);

    const changeHandle = useCallback(
      (ev: React.ChangeEvent<HTMLInputElement>) => {
        if (disabled) {
          return;
        }

        setSelfChecked(!selfChecked);

        if (onChange) {
          const event: React.ChangeEvent<HTMLInputElement> = { ...ev, target: { ...ev.target, checked: !selfChecked } };
          onChange(event);
        }
      },
      [onChange, disabled, selfChecked],
    );

    useEffect(() => {
      if (checked === undefined) {
        return;
      }
      setSelfChecked(Boolean(checked));
    }, [checked]);

    const switchRef = useRef<HTMLInputElement>(null);
    useImperativeHandle(ref, () => switchRef.current);

    return (
      <StyledSwitch className={`switch-container ${className}`} disabled={disabled} checked={selfChecked}>
        <input
          type="checkbox"
          ref={switchRef}
          onChange={changeHandle}
          checked={selfChecked}
          disabled={disabled}
          {...rest}
        />
        <span className="switch"></span>
        {children && <span className="switch-label">{children}</span>}
      </StyledSwitch>
    );
  },
);

export default Switch;
