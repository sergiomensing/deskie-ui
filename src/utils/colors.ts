export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
