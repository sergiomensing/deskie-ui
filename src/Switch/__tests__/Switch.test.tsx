import React from 'react';
import Switch from '../Switch';
import { fireEvent, render } from '../../../tests/utils';

describe('Switch', () => {
  it('should render correctly', () => {
    const { container } = render(<Switch checked={true}>test-value</Switch>);
    expect(container).toMatchSnapshot();
  });

  it('should work correctly with initial value', () => {
    const { container, rerender } = render(<Switch checked={true}>test-value</Switch>);
    expect(container.querySelector('input')?.checked).toBeTruthy();
    rerender(<Switch checked={false}>test-value</Switch>);
    expect(container.querySelector('input')?.checked).not.toBeTruthy();
  });

  it('should change value after click', () => {
    const Wrapper = () => {
      const [state, setState] = React.useState<string>('test-value1');

      return (
        <Switch data-testid="checkbox" defaultChecked onChange={() => setState('test-value2')}>
          {state}
        </Switch>
      );
    };

    const { getByTestId, getAllByText, queryAllByText } = render(<Wrapper />);

    expect(getAllByText('test-value1').length).toBe(1);

    fireEvent.click(getByTestId('checkbox'));

    expect(getAllByText('test-value2').length).toBe(1);
    expect(queryAllByText('test-value1').length).toBe(0);
  });

  it('should ignore events when disabled', () => {
    const Wrapper = () => {
      const [state, setState] = React.useState<string>('test-value1');

      return (
        <Switch data-testid="checkbox" disabled onChange={() => setState('test-value2')}>
          {state}
        </Switch>
      );
    };
    const { getByTestId, getByText, queryByText } = render(<Wrapper />);

    expect(getByText('test-value1')).toBeInTheDocument();

    fireEvent.click(getByTestId('checkbox'));

    expect(queryByText('test-value2')).toBe(null);
  });

  it('should forward ref', () => {
    const ref = React.createRef<HTMLInputElement>();
    const { container } = render(
      <Switch data-testid="checkbox" checked={true} ref={ref}>
        action
      </Switch>,
    );

    expect(container.querySelector('input')).toEqual(ref.current);
  });
});
