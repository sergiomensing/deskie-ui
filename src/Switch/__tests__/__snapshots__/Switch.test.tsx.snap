// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render correctly 1`] = `
.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #11142D;
}

.c0 input[type='checkbox'] {
  opacity: 0;
  outline: none;
  position: absolute;
  width: 0;
  height: 0;
  margin: 0;
  padding: 0;
  z-index: -1;
  background-color: transparent;
}

.c0 .switch {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c0 .switch:before {
  content: '';
  width: 36px;
  height: 24px;
  margin-right: 0.75rem;
  border-radius: 2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #D5D5E7;
  -webkit-transition: 0.1s ease;
  transition: 0.1s ease;
}

.c0 .switch:after {
  content: '';
  background-size: contain;
  background-position: center;
  width: 16px;
  height: 16px;
  left: 4px;
  top: 4px;
  position: absolute;
  border: 2px solid transparent;
  background-color: #FFFFFF;
  border-radius: 2rem;
  -webkit-transition: 0.15s ease;
  transition: 0.15s ease;
}

.c0 input:checked ~ .switch::before {
  background: #6C5DD3;
}

.c0 input:checked ~ .switch::after {
  -webkit-transform: translate3d(12px,0,0);
  -ms-transform: translate3d(12px,0,0);
  transform: translate3d(12px,0,0);
}

.c0 .switch-label {
  margin-top: 2px;
  white-space: nowrap;
}

.c0:hover .switch::before {
  background: #B2B3BD;
}

.c0:hover input:checked ~ .switch::before {
  background-color: #4535B7;
}

.c0 input:focus-visible ~ .switch:before {
  box-shadow: 0 0 0 2px #9395A1;
}

.c0 input:checked:focus-visible ~ .switch::before {
  box-shadow: 0 0 0 2px #4535B7;
}

@media screen and (min-width:1024px) {

}

<div>
  <label
    class="c0 switch-container "
  >
    <input
      type="checkbox"
    />
    <span
      class="switch"
    />
    <span
      class="switch-label"
    >
      test-value
    </span>
  </label>
</div>
`;
